### DISCUSSION TITLE
please add emit after page changed

### DISCUSSION THREAD

**Discussion by vahidalvandi** [livewire/livewire#4728] [3. Feature ideas]
https://github.com/livewire/livewire/blob/master/src/WithPagination.php
please add emit afte page changed like bellow
```

 //override WithPagination trait
    public function gotoPage($page, $pageName = 'page')
    {
        $this->setPage($page, $pageName);
        $this->emit('gotoTop');
    }

    //override WithPagination trait
    public function nextPage($pageName = 'page')
    {
        $this->setPage($this->paginators[$pageName] + 1, $pageName);
        $this->emit('gotoTop');
    }

    //override WithPagination trait
    public function previousPage($pageName = 'page')
    {
        $this->setPage(max($this->paginators[$pageName] - 1, 1), $pageName);
        $this->emit('gotoTop');
    }

    //override WithPagination trait
    public function resetPage($pageName = 'page')
    {
        $this->setPage(1, $pageName);
        $this->emit('gotoTop');
    }
```

**Comment by silnarm** [comment]
Probably would be frequently used, but I don't think adding this for everyone makes sense, when not everyone will want them, and considering how easy it is to do yourself.

Also, it would be more sensible to tap into setPage() rather than the others, or in this case it can just be wrapped, do what you want in one (and only one) place:
```php
trait WithPaginationAndPageChangedEvents
{
    use WithPagination {
        setPage as _parent_setPage;
    }

    public function setPage($page, $pageName = 'page')
    {
        $previous = $this->paginators[$page];
        $this->_parent_setPage($page, $pageName);
        $current = $this->paginators[$page];
        if ($previous != $current) {
            $this->emit('pageChanged', $pageName, $current);
            // or perhaps...
            //$this->dispatchBrowserEvent('pageChanged', ['paginator' => $pageName, 'page' => $page]);
        }
    }
}
```

**Reply by vahidalvandi** [reply]
if you add method with call it active it is ok

$model->paginate()->event();

**Comment by joshhanley** [comment]
@vahidalvandi Livewire actually has hooks for the paginators, see

https://github.com/livewire/livewire/blob/e9dd10e326e48b6f8bef99a77d9ed3d839dc74bf/src/WithPagination.php#L96-L97

So you can just setup an updated hook for your paginator and then fire the event in the hook yourself.

```php
public function updatedPage()
{
    $this->dispatchBrowserEvent('gotoTop');
}
```

Hope this helps!

**Reply by vahidalvandi** [reply]
how i can use event fire in component ?

**Reply by joshhanley** [reply]
@vahidalvandi see my answer above. Put the method `updatedPage` I wrote out above in your component. It demonstrates how to do it if you are using the default paginator name. You will need to change the hook method name if you are using a different page name.

Hope this helps!

**Reply by vahidalvandi** [reply]
@joshhanley i think updatedPage override in paginate trait

**Reply by joshhanley** [reply]
@vahidalvandi no, there is no `updatedPage` method in the pagination trait.

I wrote the implementation for this, so please listen to what I am saying. Put the `updatedPage` method in your component class, not in the pagination trait.

Trust me, try it out, it should work!

**Reply by silnarm** [reply]
My eyes **must** have passed over that very code, can't believe I missed it!
A couple of observations though @joshhanley:
1. They fire regardless of whether the page is actually changing.
2. They seem to rely on case insensitive method name matching, and also wont be consistent with other hooks (if for example, the user has snake cased paginator names) . Shouldn't the $pageName go through Str::studly() when being tacked onto $[before|after]Method ?

**Reply by joshhanley** [reply]
@silnarm hmm they should only fire when the `setPage` method is called. I just did a test locally and it only fires when the pagination has been interacted with. But doesn't on any other request. Is it doing something different for you?

And yeah you are right about the studly part, I forgot that 😅 

In Livewire's action hooks, it does it

https://github.com/livewire/livewire/blob/e9dd10e326e48b6f8bef99a77d9ed3d839dc74bf/src/ComponentConcerns/HandlesActions.php#L68-L74

So yeah it probably should be added into the paginator trait too 😬

**Reply by silnarm** [reply]
Nothing serious @joshhanley, and yes it only fires when `setPage` is called, what i was getting at is just because `setPage` has been called doesn't necessarily mean the page has actually changed. Should always be the case, and if it isn't then the user is likely at fault, so probably safe to say I am nitpicking and let's move on :-)
